### PR TITLE
Avoid application crashes & Fix issues with dynamically inserted texts in React

### PR DESCRIPTION
**Pull request by djhi**

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -85,6 +85,36 @@ export default function App() {
 }
 ```
 
+**Note**: Due to how React and [Range]() work, avoid elements that combine static texts and variables:
```
Comment (fzaninotto) [inline Readme.md]: missing link target

File: Readme.md [hunk 2]
```diff
@@ -85,6 +85,36 @@ export default function App() {
...
++        Text: <span className="content">{text}</span>
++      </div>
+    </div>
+  );
+}
+```
+
+## Troubleshooting
+
+If you get errors similar to `DOMException: Failed to execute 'setEnd' on 'Range': The offset 11 is larger than the node's length (5).`, make sure you don't mix static and dynamic texts. See the tip in previous section.
```
Comment (fzaninotto) [inline Readme.md]:
```suggestion
If you see errors in the console similar to `DOMException: Failed to execute 'setEnd' on 'Range': The offset 11 is larger than the node's length (5).`, make sure you don't mix static and dynamic texts. See the note above for an example.
```

File: src/index.js [hunk 3]
```diff
@@ -27,14 +27,18 @@ const highlightSearchTerm = ({
...
+    Array.from(elements).map((element) => {
+      getTextNodesInElementContainingText(element, search).forEach((node) => {
+        ranges.push(
+          ...getRangesForSearchTermInElement(node.parentElement, search)
+        );
+      });
     });
-  });
+  } catch (error) {
+    console.error(error);
```
Comment (fzaninotto) [inline src/index.js:40]: please add a comment explaining why it could fail, linking to the difference between `wholeText` and `textContent`

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: src/index.js [hunk 4]
```diff
@@ -57,14 +65,20 @@ const getTextNodesInElementContainingText = (element, text) => {
 const getRangesForSearchTermInElement = (element, search) => {
   const ranges = [];
   const lowerCaseSearch = search.toLowerCase();
-  if (!element.firstChild) return ranges;
-  const text = element.textContent?.toLowerCase() || "";
+  if (!element.childNodes.length === 0) return ranges;
```
Comment (fzaninotto) [inline src/index.js]: I don't understand this change. The initial `!` makes it the opposite of the previous test.

File: src/index.js [hunk 5]
```diff
@@ -57,14 +65,20 @@ const getTextNodesInElementContainingText = (element, text) => {
...
+  // In some frameworks like React, when combining static text with dynamic text, the element may have multiple Text child nodes.
+  // To avoid errors, we must find the child node that actually contains the search term.
+  const childWithSearchTerm = Array.from(element.childNodes).find((node) =>
+    node.textContent?.toLowerCase().includes(lowerCaseSearch)
+  );
+  if (!childWithSearchTerm) return ranges;
```
Comment (fzaninotto) [inline src/index.js:70]: do you still need the try/catch in the main function if you have this safety here?

**Review by fzaninotto — CHANGES_REQUESTED** [review]

File: src/index.js (hunk 4 above)
Comment (djhi) [reply src/index.js]: Fixed

File: src/index.js (hunk 5 above)
Comment (djhi) [reply src/index.js:70]: I believe we should keep it anyway

**Pull request merged**